### PR TITLE
Allow usage of implicit output for generated sources

### PR DESCRIPTION
**Pull request by mwidera**
Add option to generate source which will implicitly define output
This change is mostly related to the ninja output definition
Add function to retrieve implicit output for some of the types and also
define it on the dependent interface. Depfiles now are including also
implicit output, so e.g. headers might be passed through
Now all of them follow implicit naming scheme.
This change also aligns documentation for the generated/transform source
sections

Lastly added test to verify modified content of implicit output handling

Change-Id: I62233521ddc7b7b5058b9c45cbee62cef6f27c8d
Signed-off-by: Michal Widera <michal.widera@arm.com>

### CODE REVIEW

File: core/android_make.go [hunk 1]
```diff
@@ -648,13 +648,14 @@ func (g *androidMkGenerator) generateCommonActions(sb *strings.Builder, m *gener
 	for _, inout := range inouts {
 		if _, ok := args["headers_generated"]; ok {
 			headers := utils.Filter(inout.out, utils.IsHeader)
-			args["header_generated"] = strings.Join(headers, " ")
+			implHeaders := utils.Filter(inout.implicitOuts, utils.IsHeader)
+			args["header_generated"] = strings.Join(append(headers, implHeaders...), " ")
```
Comment (chrisdiamand) [inline core/android_make.go]: I think you can use `utils.Join` here?

File: core/build_structs.go [hunk 2]
```diff
@@ -67,6 +67,7 @@ type TargetSpecific struct {
 type dependentInterface interface {
 	phonyInterface
 	outputDir(g generatorBackend) string
+	implicitOutputs(g generatorBackend) []string
```
Comment (chrisdiamand) [inline core/build_structs.go]: Would this not make more sense in `phonyInterface` next to `outputs()`? It seems a bit strange to have different kinds of outputs in different interfaces.

(Admittedly I think both should be in `dependentInterface`, but we shouldn't change that here).

Comment (kilroyd) [reply core/build_structs.go]: I agree that it looks odd. However, I don't think implicitOutputs belongs in phonyInterface - phony support only requires outputs()

Comment (kilroyd) [reply core/build_structs.go]: Thinking about this again, it's possible to define a module that only has implicit outputs (no explicit outputs). A phony rule must therefore depend on both the explicit and implicit outputs.

So please add implicitOutputs() to the phonyInterface, and update phony target generation to include implicitOutputs(). Eventually we might want an allOutputs() function.

**Pull request merged**